addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/handler/NestBridgeHandler.java

```diff
@@ -163,144 +151,95 @@ private void refreshData() {
+    private void broadcastDevices(NestDevices devices) {
+        listeners.forEach(listener -> {
+                    if (devices.getThermostats() != null) {
+                        devices.getThermostats().values().forEach(listener::onNewNestThermostatData);
                     }
-                }
-            }
-        }
-
-        if (devices.getSmokeDetectors() != null) {
-            for (SmokeDetector smokeDetector : devices.getSmokeDetectors().values()) {
-                Thing thingSmokeDetector = getExistingDevice(smokeDetector.getDeviceId(), things);
-                if (thingSmokeDetector != null) {
-                    NestSmokeDetectorHandler handler = (NestSmokeDetectorHandler) thingSmokeDetector.getHandler();
-                    if (handler != null) {
-                        handler.updateSmokeDetector(smokeDetector);
+                    if (devices.getCameras() != null) {
+                        devices.getCameras().values().forEach(listener::onNewNestCameraData);
                     }
-                } else {
-                    for (NestDeviceAddedListener listener : listeners) {
-                        logger.debug("Found new smoke detector. {}", smokeDetector.getDeviceId());
-                        listener.onSmokeDetectorAdded(smokeDetector);
+                    if (devices.getSmokeDetectors() != null) {
+                        devices.getSmokeDetectors().values().forEach(listener::onNewNestSmokeDetectorData);
                     }
                 }
-            }
-        }
+        );
     }
 
-    private void compareStructure(Collection<Structure> structures) {
-        Bridge bridge = getThing();
-        List<Thing> things = bridge.getThings();
-
-        for (Structure struct : structures) {
-            Thing thingStructure = getExistingDevice(struct.getStructureId(), things);
-            if (thingStructure != null) {
-                NestStructureHandler handler = (NestStructureHandler) thingStructure.getHandler();
-                if (handler != null) {
-                    handler.updateStructure(struct);
-                }
-            } else {
-                for (NestDeviceAddedListener listener : listeners) {
-                    logger.debug("Found new structure {}", struct.getStructureId());
-                    listener.onStructureAdded(struct);
-                }
-            }
-        }
+    private void broadcastStructure(Collection<Structure> structures) {
+        structures.forEach(
+                structure -> listeners.forEach(l -> l.onNewNestStructureData(structure))
+        );
     }
 
-    private String buildQueryString(NestBridgeConfiguration config)
-            throws InterruptedException, TimeoutException, ExecutionException {
+    private String buildQueryString(NestBridgeConfiguration config) throws InvalidAccessTokenException {
         String stringAccessToken;
+
+        // FIXME should we ever expire access tokens?
```

Comment:
No, they should never expire, not unless someone disables them on the nest side anyway.

Reply:
I think you might be mistaken see this quote from the docs:

> Access token response
>
> A successful request returns a JSON object containing the following fields:
>
> access_token — The access token for the user. This value must be kept secure.
> expires_in — The number of seconds remaining, from the time it was requested, before the token expires.

And see the field in the existing `AccessTokenData` class, https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/data/AccessTokenData.java#L22

Reply:
I was mistaken:

> 315,360,000 Seconds =
> 3,650 Days

And yes the pin can not be re-used